SPEC.md

```diff
@@ -0,0 +1,334 @@
+# Nexus RPC HTTP Specification
+
+## Overview
+
+The Nexus protocol, as specified below, is a synchronous RPC protocol. Arbitrary length operations are modelled on top
+of a set of pre-defined synchronous RPCs.
+
+A Nexus **caller** calls a **handler**. The handler may respond inline or return a reference for a future, asynchronous
+operation. The caller can cancel an asynchronous operation, check for its outcome, or fetch its current state. The caller
+can also specify a callback URL, which the handler uses to asynchronously deliver the result of an operation when it
+is ready.
+
+## Operation Addressability
+
+An operation is addressed using three components:
+
+- [Service Name](#service-name)
+- [Operation Name](#operation-name)
+- [Operation ID](#operation-id)
+
+> **Note:** More details regarding IDs will be provided upon finalizing the ID proposal.
+
+### Service Name
+
+Service names:
+
+- MUST not be empty.
+- MUST consist of only unreserved URL characters as delineated in [RFC3986 Section 2.3][rfc3986-section-2.3].
+
+Valid characters include:
+
+```
+ALPHA / DIGIT / "-" / "." / "_" / "~"
+```
+
+### Operation Name
+
+Operation names:
+
+- MUST not be empty.
+- MUST consist of only unreserved URL characters as delineated in [RFC3986 Section 2.3][rfc3986-section-2.3].
+
+Valid characters include:
+
+```
+ALPHA / DIGIT / "-" / "." / "_" / "~"
+```
+
+### Operation ID
+
+Operation IDs:
+
+- MUST not be empty.
+- MUST consist of only unreserved URL characters as delineated in [RFC3986 Section 2.3][rfc3986-section-2.3].
+
+Valid characters include:
+
+```
+ALPHA / DIGIT / "-" / "." / "_" / "~"
+```
+
+## Schema Definitions
+
+### Failure
+
+The `Failure` object represents outcomes that are `failed` or `canceled`. The object MUST adhere to the following JSON
+schema:
+
+```yaml
+type: object
+properties:
+  message:
+    type: string
+  details:
+    type: object
+    properties:
```

Comment:
This is basically a Temporal `Payload` with slight alterations, I think it makes sense this way.

Reply:
And double-encoding of arbitrary "data" is hard to read and causes us lots of annoyance. At least https://github.com/googleapis/googleapis/blob/master/google/rpc/status.proto has `Any` "details". We should have "any details" (can be collection if needed). But I guess double encoding isn't the end of the world, just really ugly to see JSON embedded into another JSON field with escaped quotes and what not.

Reply:
I think we're going to want more of a convention here to interpret the details.

Reply:
I agree with chad. If we want to have metadata be an interprable thing, why not have it just be at the top? IE: Why not just have `message`, `metadata`, and `data` be the top three fields? What is the benefit of nesting the latter two inside `details`?

Reply:
>  just really ugly to see JSON embedded into another JSON field with escaped quotes and what not.

Agree, but we need some type of specification here so we can rehydrate these failures into language objects.

I'm wondering how often we'll see plain JSON vs. encrypted content here in practice.

So to understand the proposal from your end, you want the following, right?

```
{
  type: string (let's discuss this field separately above)
  message: string
  metadata: string -> string
  details: any (can also name this data)
}
```

So then details is any JSONable and metadata may be used to describe what goes in there...
The only problem I see is that we will want to translate this into a Temporal payload and we would need some convention to know whether details is a string that encodes binary data or should be interpreted as a plain string.
So maybe reserve a metadata field to express that?

Reply:
Another option would be to have separate mutually exclusive fields for binary vs. non binary data.

Reply:
I like that option. Binary data can say how it is encoded in the metadata, and non-binary data can just be json. Does it even need to be separate fields? I mean I suppose _technically_ you could have raw binary that happens to also be valid json but that seems so unlikely as to be unimportant.

Reply:
I think the idea here is that JSON is not encoded into a string, it would be inlined.
Then if details is a string, it may represent either a plain string or some encoded data.
We need to know whether to treat it as a string or decode it.

Reply:
> I think the idea here is that JSON is not encoded into a string, it would be inlined.

:+1: Same as an actual string, hence the "any" not "any object". But I say don't worry about binary. It should _always_ be some JSON primitive, and if they want binary, they encode (we'd have to anyways).

Reply:
I just want this spec to include the method by which we tell binary data encoded as string from plain string.

I'm going with

```
{
  "metadata": {
     "Content-Transfer-Encoding": "base64",
  },
  "details": "<base64-encoded-string>",
}
```

So if details is a string and metadata["Content-Transfer-Encoding"] is `base64` we'll consider details as bytes.

Reply:
I just realized that message may be encoded too.

New proposal:

```
{
  message: string | Payload // may be a payload
  details: any // may be a Payload
}
```

Where `Payload` is:
```
{
  metadata: map[string]string
  data: base64 encoded string
}
```

And when transferred (either via request or response) headers should include:
```
-- Option 1 --
Content-Type: application/nexus-payload-failure+json
-- Option 2 --
Nexus-Failure-Encoding: payloads | none
```

Reply:
Thought about it some more and encrypting the message can be done by encrypting the entire body and providing the appropriate http headers.

Reply:
Now debating whether the framework should support binary details.

Reply:
Just keep it basic "any JSON" and let the user figure out how to represent binary since it's not a JSON type (so base64 string is probably most common for them, but protocol shouldn't care).